File: lib/domain/farm/interactors/start_farm_weather_data_fetch_interactor.rb
# frozen_string_literal: true

module Domain
  module Farm
    module Interactors
      class StartFarmWeatherDataFetchInteractor
        def initialize(farm_gateway:, fetch_weather_data_enqueue_port:)
          @farm_gateway = farm_gateway
          @fetch_weather_data_enqueue_port = fetch_weather_data_enqueue_port
        end

        def call(input_dto)
          farm = @farm_gateway.find_by_id(input_dto.farm_id)
          return nil unless farm.has_coordinates?

          as_of = input_dto.as_of
          attrs = Calculators::FarmWeatherProgressCalculator.start_fetch_attrs(as_of: as_of)
          @farm_gateway.update_weather_progress(input_dto.farm_id, attrs)

          blocks = Calculators::FarmWeatherProgressCalculator.weather_fetch_blocks(as_of: as_of)
          @fetch_weather_data_enqueue_port.enqueue_farm_weather_fetch(
            farm_id: farm.id,
            latitude: farm.latitude,
            longitude: farm.longitude,
            blocks: blocks
          )
          farm
        end
      end
    end
  end
end

File: crates/agrr-domain/src/farm/interactors/start_farm_weather_data_fetch_interactor.rs
//! Ruby: `Domain::Farm::Interactors::StartFarmWeatherDataFetchInteractor`

use crate::farm::calculators::FarmWeatherProgressCalculator;
use crate::farm::dtos::StartFarmWeatherDataFetchInput;
use crate::farm::entities::FarmEntity;
use crate::farm::gateways::FarmGateway;
use crate::shared::ports::FetchWeatherDataEnqueuePort;

pub struct StartFarmWeatherDataFetchInteractor<'a, G, E> {
    farm_gateway: &'a G,
    fetch_weather_data_enqueue_port: &'a E,
}

impl<'a, G, E> StartFarmWeatherDataFetchInteractor<'a, G, E>
where
    G: FarmGateway,
    E: FetchWeatherDataEnqueuePort,
{
    pub fn new(farm_gateway: &'a G, fetch_weather_data_enqueue_port: &'a E) -> Self {
        Self {
            farm_gateway,
            fetch_weather_data_enqueue_port,
        }
    }

    pub fn call(
        &self,
        input: StartFarmWeatherDataFetchInput,
    ) -> Result<Option<FarmEntity>, Box<dyn std::error::Error + Send + Sync>> {
        let farm = self.farm_gateway.find_by_id(input.farm_id)?;
        if !farm.has_coordinates() {
            return Ok(None);
        }

        let as_of_year = input.as_of.year();
        let attrs = FarmWeatherProgressCalculator::start_fetch_attrs(as_of_year);
        self.farm_gateway
            .update_weather_progress(input.farm_id, attrs)?;

        let blocks = FarmWeatherProgressCalculator::weather_fetch_date_blocks(input.as_of);
        let (lat, lon) = farm.coordinates();
        self.fetch_weather_data_enqueue_port.enqueue_farm_weather_fetch(
            farm.id,
            lat.unwrap_or(0.0),
            lon.unwrap_or(0.0),
            &blocks,
        );

        Ok(Some(farm))
    }
}

#[cfg(test)]
mod interactors_start_farm_weather_data_fetch_interactor_test_inline {
    use super::*;
    include!(concat!(env!("CARGO_MANIFEST_DIR"), "/test/farm/interactors_start_farm_weather_data_fetch_interactor_test.rs"));
}
